### ISSUE TITLE
ActiveSupport::Duration bug regarding granularity with higher numbers

### ISSUE THREAD

**Issue by muja** [rails/rails#31633]
### Steps to reproduce

    $ irb
    > require 'active_support/duration'
    true
    > seconds = 3155695199
    > Active::Support.build(seconds)
    99 years, 11 months, 4 weeks, 2 days, 10 hours, 29 minutes, and 5.0 seconds
    > Active::Support.build(seconds + 1)
    100 years

### Expected behavior
The first line should either be
`99 years 11 months 4 weeks 2 days 23 hours 59 minutes 59.0 seconds`
or the second line should be 
`99 years, 11 months, 4 weeks, 2 days, 10 hours, 29 minutes, and 6.0 seconds`

### Actual behavior
Removing one second from 100.years removes ~14 hours

### System configuration
**Rails version**:
5.1.4 and master
**Ruby version**:
2.3.1

**Comment by owaiswiz** [comment]
Hello,
The exact no of seconds in a gregorian calendar is 31556952 because the mean length is 365.2425 days - https://en.wikipedia.org/wiki/Gregorian_calendar
So 3155695199 + 1 ( 3155695200) should in fact equal 100 years. 

Hence you don't get the output you expect because it doesn't derive the value of a year from the rounded 365 days.

**Comment by muja** [comment]
I know that. Thanks to your example, I realized just know that what is displayed in the output equals to a whole 365 days + 10 hours etc. Why doesn't it say "12 months 10 hours 29 minutes 5.0 seconds"?

Okay, I realized now that the months are fragmented as well and `1.month != 30.days`. So the when I display "1 month" to my end user, what I am actually saying is 30 days, 10 hours, 29, 5 seconds? Seems off, but I guess there was a reason for this choice.
